### PR TITLE
[interpreter] Add command-line option to set max recursion depth

### LINKED ISSUE #1200: [interpreter] Add command line option to control maximum recursion depth
At this moment, the maximum recursion depth is [hard-coded to 300](https://github.com/WebAssembly/spec/blob/a7a1856cc8e0e9cb6378acb083bb9c43494d05b1/interpreter/exec/eval.ml#L68). When experimenting with recursive functions, I rapidly reached this limit and had to change manually it in the source code before recompiling the interpreter.

This number feels arbitrary and I cannot think of any reason to prevent users from setting it to a number that fit their purpose. I propose to add a command line option (e.g. `-max-recursion-depth <natural-number>`) with 300 being the default value.

I am just starting to get to know the code base but, if this suggestion is accepted, I am willing to provide a pull request.

### PR DESCRIPTION
**Pull request by mdesharnais**
In Eval.step function, fix Invoke case which did not handle negative integers.

close #1200

### CODE REVIEW

File: interpreter/exec/eval.ml [hunk 1]
```diff
@@ -314,7 +318,7 @@ let rec step (c : config) : config =
       let c' = step {frame = frame'; code = code'; budget = c.budget - 1} in
       vs, [Frame (n, c'.frame, c'.code) @@ e.at]
 
-    | Invoke func, vs when c.budget = 0 ->
+    | Invoke func, vs when c.budget <= 0 ->
```
Comment (rossberg) [inline interpreter/exec/eval.ml:321]:
```suggestion
    | Invoke func, vs when c.budget = 0 ->
```

File: interpreter/main/main.ml [hunk 2]
```diff
@@ -31,7 +31,10 @@ let argspec = Arg.align
   "-h", Arg.Clear Flags.harness, " exclude harness for JS conversion";
   "-d", Arg.Set Flags.dry, " dry, do not run program";
   "-t", Arg.Set Flags.trace, " trace execution";
-  "-v", Arg.Unit banner, " show version"
+  "-v", Arg.Unit banner, " show version";
+  "-max-recursion-depth", Arg.Set_int Flags.max_recursion_depth,
```
Comment (rossberg) [inline interpreter/main/main.ml:35]:
```suggestion
  "-r", Arg.Set_int Flags.max_recursion_depth,
```

File: interpreter/exec/eval.ml [hunk 3]
```diff
@@ -65,7 +65,11 @@ type config =
 }
 
 let frame inst locals = {inst; locals}
-let config inst vs es = {frame = frame inst []; code = vs, es; budget = 300}
+let config inst vs es = {
+  frame = frame inst [];
+  code = vs, es;
+  budget = !Flags.max_recursion_depth
+}
```
Comment (rossberg) [inline interpreter/exec/eval.ml:72]:
```suggestion
let config inst vs es =
  { frame = frame inst [];
    code = vs, es;
    budget = max 0 !Flags.max_recursion_depth
  }
```

**Comment by mdesharnais** [comment]
Hi, just a friendly remainder. I integrated all your suggestions and added a short explanation in the README. Is there anything else I should change?

**Comment by ngzhian** [comment]
@mdesharnais Can you follow the instructions https://labs.w3.org/repo-manager/pr/id/WebAssembly/spec/1201 to satisfy the ipr check? @dtig @dschuff fyi chairs to help guide this.

**Review by rossberg — COMMENTED** [review]
Oops, sorry, I somehow managed to completely miss the update on this PR at the time, not sure how!

Thanks for the changes! One further change I'd suggest: the flag adjusts the maximum call depth. This isn't tied to recursion, so calling it that is a bit misleading. Can we call it `max_call_depth` instead?

**Comment by rossberg** [comment]
Closing, since this has been implemented in the meantime.

**Pull request closed without merging**